src/v/cluster/controller_stm.cc

```diff
@@ -0,0 +1,146 @@
+    });
+}
+
+ss::future<> controller_stm::write_snapshot() {
+    if (!_feature_table.is_active(features::feature::controller_snapshots)) {
+        vlog(clusterlog.warn, "skipping snapshotting, feature not enabled");
+        co_return;
+    }
+    co_await _write_snapshot_mtx.with([this] { return do_write_snapshot(); });
+}
+
+struct controller_stm::snapshot {
+    model::offset offset;
+    controller_snapshot data;
+};
+
+ss::future<> controller_stm::do_write_snapshot() {
+    auto snapshot = co_await _apply_mtx.with(
+      [this] { return do_make_snapshot(); });
+
+    if (!snapshot) {
+        co_return;
+    }
+
+    iobuf snapshot_buf = serde::to_iobuf(std::move(snapshot->data));
+    co_await _raft->write_snapshot(
+      raft::write_snapshot_cfg(snapshot->offset, std::move(snapshot_buf)));
+}
+
+ss::future<std::optional<controller_stm::snapshot>>
+controller_stm::do_make_snapshot() {
+    if (_raft->last_snapshot_index() >= last_applied()) {
+        co_return std::nullopt;
+    }
+
+    controller_snapshot data;
+    co_await std::apply(
+      [&data](auto&&... stms) {
+          return ss::when_all(stms.fill_snapshot(data)...);
+      },
+      _state);
+
+    vlog(clusterlog.info, "created snapshot at offset {}", last_applied());
+    co_return snapshot{.offset = last_applied(), .data = std::move(data)};
+}
+
+ss::future<model::offset> controller_stm::load_snapshot() {
+    vlog(
+      clusterlog.info,
+      "loading snapshot, previous last_applied: {}",
+      last_applied());
+
+    auto reader = co_await _raft->get_snapshot_reader();
+    if (!reader) {
+        throw std::runtime_error{fmt_with_ctx(
+          fmt::format,
+          "encountered a gap in the raft log (last_applied: {}, log start "
+          "offset: {}), but can't find the snapshot",
+          last_applied(),
+          _raft->start_offset())};
+    }
+
+    auto snapshot_offset = co_await do_load_snapshot(*reader).finally(
+      [&reader] { return reader->close(); });
+    co_return snapshot_offset;
+}
+
+ss::future<model::offset>
+controller_stm::do_load_snapshot(storage::snapshot_reader& reader) {
+    auto md_parser = iobuf_parser(co_await reader.read_metadata());
+    auto metadata = reflection::adl<raft::snapshot_metadata>{}.from(md_parser);
+
+    const size_t size = co_await reader.get_snapshot_size();
+    vlog(
+      clusterlog.info,
+      "loading snapshot at {}, size: {}",
+      metadata.last_included_index,
+      size);
+
+    iobuf snap_buf = co_await read_iobuf_exactly(reader.input(), size);
+    auto snapshot = serde::from_iobuf<controller_snapshot>(std::move(snap_buf));
+
+    try {
+        co_await std::apply(
+          [&metadata, &snapshot](auto&&... stms) {
+              return ss::when_all(stms.apply_snapshot(metadata, snapshot)...);
+          },
+          _state);
+    } catch (...) {
+        vassert(
+          false,
+          "Failed to apply snapshot: {}. "
+          "State inconsistency possible, aborting.",
```

Comment:
Should we log the snapshot path here? Think the only way to recover from this situation is to remove the snapshot path from disk.

Reply:
location of the controller is kind of well-known, but anyway won't hurt, so added